swiper.el

```diff
@@ -193,49 +199,85 @@ If the input is empty, select the previous history element instead."
-                     (groups (number-sequence 1 ivy--subexps))
-                     (default
-                      (list
-                       (mapconcat (lambda (i) (format "\\%d" i)) groups " ")
-                       (format "\\,(concat %s)"
-                               (if (<= ivy--subexps 1)
-                                   "\\&"
-                                 (mapconcat
-                                  (lambda (i) (format "\\%d" i))
-                                  groups
-                                  " \" \" ")))))
-                     (to
-                      (query-replace-compile-replacement
-                       (ivy-read
-                        (format "Query replace %s with: " from) nil
-                        :def default
-                        :caller 'swiper-query-replace)
-                       t)))
-                (swiper--cleanup)
-                (ivy-exit-with-action
-                 (lambda (_)
-                   (with-ivy-window
-                     (move-beginning-of-line 1)
-                     (let ((inhibit-read-only t))
-                       (perform-replace from to
-                                        t t nil))))))
-           (swiper--query-replace-cleanup)))))
-
-(ivy-configure 'swiper-query-replace
-  :update-fn #'swiper--query-replace-updatefn)
-(put 'swiper-query-replace 'no-counsel-M-x t)
+;;;###autoload
+(defun swiper-query-replace (&optional initial-regexp initial-replacement start end)
+  "Read a regexp with Swiper and start a query replace operation.
+INITIAL-REGEXP and INITIAL-REPLACEMENT are the initial inputs of
+the respective prompts.  When called interactively with a prefix
+argument, the values of the previous query replace are used.
+
+START and END specify the region to operate on.  When called
+interactively from outside the minibuffer, the current region is
+used if active.
+
+This command can also be called from within a Swiper session.  In
+this case, it doesn't ask for a regexp, and instead uses the
+current Swiper input."
+  (interactive (list (when current-prefix-arg (caar query-replace-defaults))
+                     (when current-prefix-arg (cdar query-replace-defaults))
+                     (when (use-region-p) (region-beginning))
+                     (when (use-region-p) (region-end))))
+  (barf-if-buffer-read-only)
+  (if (window-minibuffer-p)
+      (ivy-exit-with-action
+       (lambda (x)
+         (swiper-isearch-action x)
+         (swiper-query-replace t)))
+    (when start
+      (deactivate-mark) ;; TODO: maybe add a fake active region as overlay
+      (goto-char start))
+    (when (string-or-null-p initial-regexp)
+      (let ((ivy-fixed-height-minibuffer t)
+            (cursor-in-non-selected-windows nil)
+            (swiper-min-highlight 1))
+        (ivy-read
+         "Swiper query replace: "
+         #'swiper-isearch-function
+         :initial-input initial-regexp
+         :keymap swiper-isearch-map
```

Comment:
This gives a "reference to free variable `swiper-isearch-map`" byte-compiler warning, which probably means you need to define that variable before this function.

Reply:
I've replaced this with `swiper-map`. Then `M-n` will not quote the query, but at least no large reorganizing of the code is needed.

Reply:
I don't know which is better, but FWIW moving a `defvar` verbatim doesn't qualify as "large reorganizing of the code".

Reply:
True; but then a random piece of swiper-isearch will be outside of the corresponding `;;*`-section of the file. I was assuming this is undesired.

Reply:
Rules are made to be broken ;).

Reply:
Or rather, there are always exceptions that prove the rule.